src/network/network_survey.cpp

```diff
@@ -0,0 +1,400 @@
+		if (c->ai_info == nullptr) {
+			company["type"] = "human";
+		} else {
+			company["type"] = "ai";
+			company["script"] = fmt::format("{}.{}", c->ai_info->GetName(), c->ai_info->GetVersion());
+		}
+
+		for (VehicleType type = VEH_BEGIN; type < VEH_COMPANY_END; type++) {
+			uint amount = c->group_all[type].num_vehicle;
+			company["vehicles"][_vehicle_type_to_string[type]] = amount;
+		}
+
+		company["infrastructure"]["road"] = c->infrastructure.GetRoadTotal();
+		company["infrastructure"]["tram"] = c->infrastructure.GetTramTotal();
+		company["infrastructure"]["rail"] = c->infrastructure.GetRailTotal();
+		company["infrastructure"]["signal"] = c->infrastructure.signal;
+		company["infrastructure"]["water"] = c->infrastructure.water;
+		company["infrastructure"]["station"] = c->infrastructure.station;
+		company["infrastructure"]["airport"] = c->infrastructure.airport;
+	}
+}
+
+/**
+ * Convert GRF information to JSON.
+ *
+ * @param survey The JSON object.
+ */
+static void SurveyGrfs(nlohmann::json &survey)
+{
+	for (GRFConfig *c = _grfconfig; c != nullptr; c = c->next) {
+		auto grfid = fmt::format("{:08x}", BSWAP32(c->ident.grfid));
+		auto &grf = survey[grfid];
+
+		grf["md5sum"] = MD5SumToString(c->ident.md5sum);
+		grf["status"] = c->status;
+
+		if ((c->palette & GRFP_GRF_MASK) == GRFP_GRF_UNSET) grf["palette"] = "unset";
+		if ((c->palette & GRFP_GRF_MASK) == GRFP_GRF_DOS) grf["palette"] = "dos";
+		if ((c->palette & GRFP_GRF_MASK) == GRFP_GRF_WINDOWS) grf["palette"] = "windows";
+		if ((c->palette & GRFP_GRF_MASK) == GRFP_GRF_ANY) grf["palette"] = "any";
+
+		if ((c->palette & GRFP_BLT_MASK) == GRFP_BLT_UNSET) grf["blitter"] = "unset";
+		if ((c->palette & GRFP_BLT_MASK) == GRFP_BLT_32BPP) grf["blitter"] = "32bpp";
+
+		grf["is_static"] = HasBit(c->flags, GCF_STATIC);
+
+		std::vector<uint32> parameters;
+		for (int i = 0; i < c->num_params; i++) {
+			parameters.push_back(c->param[i]);
+		}
+		grf["parameters"] = parameters;
+	}
+}
+
+/**
+ * Convert game-script information to JSON.
+ *
+ * @param survey The JSON object.
+ */
+static void SurveyGameScript(nlohmann::json &survey)
+{
+	if (Game::GetInfo() == nullptr) return;
+
+	survey = fmt::format("{}.{}", Game::GetInfo()->GetName(), Game::GetInfo()->GetVersion());
+}
+
+#endif /* WITH_NLOHMANN_JSON */
+
+/**
+ * Create the payload for the survey.
+ *
+ * @param reason The reason for sending the survey.
+ * @param for_preview Whether the payload is meant for preview. This indents the result, and filters out the id/key.
+ * @return std::string The JSON payload as string for the survey.
+ */
+std::string NetworkSurveyHandler::CreatePayload(Reason reason, bool for_preview)
+{
+#ifndef WITH_NLOHMANN_JSON
+	return "";
+#else
+	nlohmann::json survey;
+
+	survey["schema"] = NETWORK_SURVEY_VERSION;
+	survey["reason"] = reason;
+
+	/* id and key are used to authenticate surveys results and are not meant
+	 * to be previewed by the user. */
+	if (!for_preview) {
```

Comment:
Hiding things you are going to send doesn't feel right to me, if you add the UI to preview it to be open and clear about what it about to be sent.
Side note: do we mention the URL this gets uploaded to, and should we maybe show that as well in the preview UI?

Reply:
Security vs transparency is always a battle. I have no problems making the `id` visible, but they `key` I rather don't show. But I can show: `key: <censored>`, if that works for you?

Of course it is trivial for anyone with knowledge to figure out the key anyway, but at least it makes it more difficult for the other 90% of the people :)

Reply:
![image](https://github.com/OpenTTD/OpenTTD/assets/1663690/4abde722-42a8-43b6-85cf-5364b541f609)

ID visible, key redacted. Good enough?